<a href="https://colab.research.google.com/github/SergeiVKalinin/MSE_Fall_2025/blob/main/Module%202/5_Matminer_bulk_modulus.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The notebook from the original matminer repository:

https://github.com/hackingmaterials/matminer

# Matminer introduction - Predicting bulk modulus

This notebook demonstrates some of the basic features of matminer.

This notebook was last updated 06/07/21 for version 0.7.0 of matminer and verison 0.0.2 of figrecipes (which you can download by cloning the [figrecipes source repo](https://github.com/hackingmaterials/figrecipes)).


**Note that in order to get the in-line plotting to work, you might need to start Jupyter notebook with a higher data rate, e.g., ``jupyter notebook --NotebookApp.iopub_data_rate_limit=1.0e10``. We recommend you do this before starting.**


## Overview
In this notebook, we will:
1. Load and examine a dataset in a pandas dataframe
2. Add descriptors to the dataframe using matminer
3. Train, compare, and visualize several machine learning methods with scikit-learn and matminer FigRecipes.



## 1. Load and process data set

Matminer comes pre-loaded with several example data sets you can use. Below, we'll load a data set of computed elastic properties of materials which is sourced from the paper:  "Charting the complete elastic properties of inorganic crystalline compounds", M. de Jong *et al.*, Sci. Data. 2 (2015) 150009.

In [1]:
!pip install matminer
!pip install mdf_forge
!pip -q install mdf-toolbox mdf-forge
!pip install figrecipes

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 55.6/55.6 kB 2.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 45.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 51.9/51.9 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 37.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.7/1.7 MB 33.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 313.6/313.6 kB 16.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 332.3/332.3 kB 17.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 119.7/119.7 kB 9.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 809.1/809.1 kB 28.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.1/60.1 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 754.1/754.1 kB 30.2 MB/s eta 0:00:00
  Created wheel for bibtexparser: filename=bibtexparser-1.4.3-py3-none-any.whl size=4354

In [2]:
from matminer.datasets.convenience_loaders import load_elastic_tensor
df = load_elastic_tensor()  # loads dataset in a pandas DataFrame object

Fetching elastic_tensor_2015.json.gz from https://ndownloader.figshare.com/files/13220603 to /usr/local/lib/python3.12/dist-packages/matminer/datasets/elastic_tensor_2015.json.gz


Fetching https://ndownloader.figshare.com/files/13220603 in MB: 1.1182079999999999MB [00:00, 229.15MB/s]                


## 1.1 A first look at the data set

The data set comes as a pandas DataFrame, which is a kind of "spreadsheet" object in Python. DataFrames have several useful methods you can use to explore and clean the data, some of which we'll explore below.

In [3]:
df.head()

material_id    formula  nsites  space_group      volume  \
0    mp-10003    Nb4CoSi      12          124  194.419802   
1    mp-10010  Al(CoSi)2       5          164   61.987320   
2    mp-10015       SiOs       2          221   25.952539   
3    mp-10021         Ga       4           63   76.721433   
4    mp-10025      SiRu2      12           62  160.300999   

                                           structure  elastic_anisotropy  \
0  [[0.94814328 2.07280467 2.5112    ] Nb, [5.273...            0.030688   
1  [[0. 0. 0.] Al, [1.96639263 1.13529553 0.75278...            0.266910   
2   [[1.480346 1.480346 1.480346] Si, [0. 0. 0.] Os]            0.756489   
3  [[0.         1.09045794 0.84078375] Ga, [0.   ...            2.376805   
4  [[1.0094265  4.24771709 2.9955487 ] Si, [3.028...            0.196930   

      G_Reuss       G_VRH     G_Voigt     K_Reuss       K_VRH     K_Voigt  \
0   96.844535   97.141604   97.438674  194.267623  194.268884  194.270146   
1   93.939650   96.252006   98.564362  173.647763  175.449907  177.252050   
2  120.962289  130.112955  139.263621  295.077545  295.077545  295.077545   
3   12.205989   15.101901   17.997812   49.025963   49.130670   49.235377   
4  100.110773  101.947798  103.784823  255.055257  256.768081  258.480904   

   poisson_ratio                                  compliance_tensor  \
0       0.285701  [[0.004385293093993, -0.0016070693558990002, -...   
1       0.268105  [[0.0037715428949660003, -0.000844229828709, -...   
2       0.307780  [[0.0019959391925840004, -0.000433146670736000...   
3       0.360593  [[0.021647143908635, -0.005207263618160001, -0...   
4       0.324682  [[0.00410214297725, -0.001272204332729, -0.001...   

                                      elastic_tensor  \
0  [[311.33514638650246, 144.45092552856926, 126....   
1  [[306.93357350984974, 88.02634955100905, 105.6...   
2  [[569.5291276937579, 157.8517489654999, 157.85...   
3  [[69.28798774976904, 34.7875015216915, 37.3877...   
4  [[349.3767766177825, 186.67131003104407, 176.4...   

                             elastic_tensor_original  
0  [[311.33514638650246, 144.45092552856926, 126....  
1  [[306.93357350984974, 88.02634955100905, 105.6...  
2  [[569.5291276937579, 157.8517489654999, 157.85...  
3  [[70.13259066665267, 40.60474945058445, 37.387...  
4  [[407.4791016459293, 176.4759188081947, 213.83...

## 1.2 Removing unneeded columns from the data set

The data set above has many columns - we won't need all this data for our modeling. We'll mainly be trying to predict ``K_VRH`` and ``G_VRH`` (the Voight-Reuss-Hill average of the bulk and shear modulus, respectively) and the ``elastic_anisotropy``. We can drop most of the other output data. We also don't need various metadata such as the ``cif`` string of the structure since the crystal structure is already embedded in the ``structure`` column.

In [4]:
df.columns

Index(['material_id', 'formula', 'nsites', 'space_group', 'volume',
       'structure', 'elastic_anisotropy', 'G_Reuss', 'G_VRH', 'G_Voigt',
       'K_Reuss', 'K_VRH', 'K_Voigt', 'poisson_ratio', 'compliance_tensor',
       'elastic_tensor', 'elastic_tensor_original'],
      dtype='object')

In [5]:
unwanted_columns = ["volume", "nsites", "compliance_tensor", "elastic_tensor",
                    "elastic_tensor_original", "K_Voigt", "G_Voigt", "K_Reuss", "G_Reuss"]
df = df.drop(unwanted_columns, axis=1)

In [6]:
df.head()

material_id    formula  space_group  \
0    mp-10003    Nb4CoSi          124   
1    mp-10010  Al(CoSi)2          164   
2    mp-10015       SiOs          221   
3    mp-10021         Ga           63   
4    mp-10025      SiRu2           62   

                                           structure  elastic_anisotropy  \
0  [[0.94814328 2.07280467 2.5112    ] Nb, [5.273...            0.030688   
1  [[0. 0. 0.] Al, [1.96639263 1.13529553 0.75278...            0.266910   
2   [[1.480346 1.480346 1.480346] Si, [0. 0. 0.] Os]            0.756489   
3  [[0.         1.09045794 0.84078375] Ga, [0.   ...            2.376805   
4  [[1.0094265  4.24771709 2.9955487 ] Si, [3.028...            0.196930   

        G_VRH       K_VRH  poisson_ratio  
0   97.141604  194.268884       0.285701  
1   96.252006  175.449907       0.268105  
2  130.112955  295.077545       0.307780  
3   15.101901   49.130670       0.360593  
4  101.947798  256.768081       0.324682

## 1.3 Getting a feel for the data using descriptive statistics

A pandas DataFrame includes a function called ``describe()`` that helps determine statistics for the various numerical / categorical columns in the data.

In [7]:
df.describe()

space_group  elastic_anisotropy        G_VRH        K_VRH  \
count  1181.000000         1181.000000  1181.000000  1181.000000   
mean    163.403895            2.145013    67.543145   136.259661   
std      65.040733           19.140097    44.579408    72.886978   
min       4.000000            0.000005     2.722175     6.476135   
25%     124.000000            0.145030    34.117959    76.435350   
50%     193.000000            0.355287    59.735163   130.382766   
75%     221.000000            0.923117    91.332142   189.574194   
max     229.000000          397.297866   522.921225   435.661487   

       poisson_ratio  
count    1181.000000  
mean        0.287401  
std         0.062177  
min         0.042582  
25%         0.249159  
50%         0.290198  
75%         0.328808  
max         0.467523

Sometimes, the ``describe()`` function will reveal outliers that indicate mistakes in the data. For example, negative hence unphysical minimum bulk/shear moduli or maximum bulk/shear moduli that are too high. In this case, the data looks ok at first glance; meaing that there are no clear problems with the ranges of the various properties. Therefore, and we won't filter out any data.

Note that the ``describe()`` function only describes numerical columns by default.

# 2. Add descriptors to the data ("featurization")

We are seeking to find relationships between the inputs (composition and crystal structure of a material) and outputs (elastic properties such as ``K_VRH``, ``G_VRH``, and ``elastic_anisotropy``). To find such relations, we need to "featurize" the input data such that they are numbers that meaningfully represent the underlying physical quantity. For example, one "feature" or "descriptor" of the composition of a material such as ``Nb4CoSi`` would be the standard deviation of the Pauling electronegativity of the elements in the compound (weighted by stoichiometry). Compositions with high values of this quantity would be more ionic and quantities with lower values would tend towards covalent or ionic. A descriptor of the crystal structure might be the average coordination number of sites; higher coordination numbers indicate more bonds and therefore might indicate stiffer materials. With matminer, we can start generating hundreds of possible descriptors using the descriptor library that is available. Data mining techniques can help narrow down which descriptors are the most relevant to the target problem using the available output data as a guide.

## 2.1 Add composition-based features

A major class of featurizers available in matminer uses the chemical composition to featurize the input data. Let's add some composition based features to our DataFrame.

The first step is to have a column representing the chemical composition as a pymatgen Composition object. One way to do this is to use the conversions Featurizers in matminer to turn a String composition (our ``formula`` column from before) into a pymatgen Composition.

In [8]:
from matminer.featurizers.conversions import StrToComposition
df = StrToComposition().featurize_dataframe(df, "formula")
df.head()

StrToComposition:   0%|          | 0/1181 [00:00<?, ?it/s]

material_id    formula  space_group  \
0    mp-10003    Nb4CoSi          124   
1    mp-10010  Al(CoSi)2          164   
2    mp-10015       SiOs          221   
3    mp-10021         Ga           63   
4    mp-10025      SiRu2           62   

                                           structure  elastic_anisotropy  \
0  [[0.94814328 2.07280467 2.5112    ] Nb, [5.273...            0.030688   
1  [[0. 0. 0.] Al, [1.96639263 1.13529553 0.75278...            0.266910   
2   [[1.480346 1.480346 1.480346] Si, [0. 0. 0.] Os]            0.756489   
3  [[0.         1.09045794 0.84078375] Ga, [0.   ...            2.376805   
4  [[1.0094265  4.24771709 2.9955487 ] Si, [3.028...            0.196930   

        G_VRH       K_VRH  poisson_ratio   composition  
0   97.141604  194.268884       0.285701  (Nb, Co, Si)  
1   96.252006  175.449907       0.268105  (Al, Co, Si)  
2  130.112955  295.077545       0.307780      (Si, Os)  
3   15.101901   49.130670       0.360593          (Ga)  
4  101.947798  256.768081       0.324682      (Si, Ru)

As you can see, we now have a new Composition column above. The visualization in the DataFrame is not so clear, but the column embeds both the elements *and* the amounts of each element in the composition (not shown).

Next, we'll use one of the featurizers in matminer to add a suite of descriptors to the DataFrame.

In [9]:
from matminer.featurizers.composition import ElementProperty

ep_feat = ElementProperty.from_preset(preset_name="magpie")
df = ep_feat.featurize_dataframe(df, col_id="composition")  # input the "composition" column to the featurizer
df.head()

/usr/local/lib/python3.12/dist-packages/matminer/utils/data.py:326: UserWarning: MagpieData(impute_nan=False):
In a future release, impute_nan will be set to True by default.
                    This means that features that are missing or are NaNs for elements
                    from the data source will be replaced by the average of that value
                    over the available elements.
                    This avoids NaNs after featurization that are often replaced by
                    dataset-dependent averages.
  warnings.warn(f"{self.__class__.__name__}(impute_nan=False):\n" + IMPUTE_NAN_WARNING)


ElementProperty:   0%|          | 0/1181 [00:00<?, ?it/s]

material_id    formula  space_group  \
0    mp-10003    Nb4CoSi          124   
1    mp-10010  Al(CoSi)2          164   
2    mp-10015       SiOs          221   
3    mp-10021         Ga           63   
4    mp-10025      SiRu2           62   

                                           structure  elastic_anisotropy  \
0  [[0.94814328 2.07280467 2.5112    ] Nb, [5.273...            0.030688   
1  [[0. 0. 0.] Al, [1.96639263 1.13529553 0.75278...            0.266910   
2   [[1.480346 1.480346 1.480346] Si, [0. 0. 0.] Os]            0.756489   
3  [[0.         1.09045794 0.84078375] Ga, [0.   ...            2.376805   
4  [[1.0094265  4.24771709 2.9955487 ] Si, [3.028...            0.196930   

        G_VRH       K_VRH  poisson_ratio   composition  \
0   97.141604  194.268884       0.285701  (Nb, Co, Si)   
1   96.252006  175.449907       0.268105  (Al, Co, Si)   
2  130.112955  295.077545       0.307780      (Si, Os)   
3   15.101901   49.130670       0.360593          (Ga)   
4  101.947798  256.768081       0.324682      (Si, Ru)   

   MagpieData minimum Number  ...  MagpieData range GSmagmom  \
0                       14.0  ...                   1.548471   
1                       13.0  ...                   1.548471   
2                       14.0  ...                   0.000000   
3                       31.0  ...                   0.000000   
4                       14.0  ...                   0.000000   

   MagpieData mean GSmagmom  MagpieData avg_dev GSmagmom  \
0                  0.258079                     0.430131   
1                  0.619388                     0.743266   
2                  0.000000                     0.000000   
3                  0.000000                     0.000000   
4                  0.000000                     0.000000   

   MagpieData mode GSmagmom  MagpieData minimum SpaceGroupNumber  \
0                       0.0                                194.0   
1                       0.0                                194.0   
2                       0.0                                194.0   
3                       0.0                                 64.0   
4                       0.0                                194.0   

   MagpieData maximum SpaceGroupNumber  MagpieData range SpaceGroupNumber  \
0                                229.0                               35.0   
1                                227.0                               33.0   
2                                227.0                               33.0   
3                                 64.0                                0.0   
4                                227.0                               33.0   

   MagpieData mean SpaceGroupNumber  MagpieData avg_dev SpaceGroupNumber  \
0                        222.833333                             9.611111   
1                        213.400000                            15.520000   
2                        210.500000                            16.500000   
3                         64.000000                             0.000000   
4                        205.000000                            14.666667   

   MagpieData mode SpaceGroupNumber  
0                             229.0  
1                             194.0  
2                             194.0  
3                              64.0  
4                             194.0  

[5 rows x 141 columns]

As you can see, we now have many more columns in the DataFrame (a total of 141 columns - not all are shown). So we added *many* features to our data!

As an aside, note that each featurizer also has a ``citations()`` function that tells you where to find out more about the Featurizer.

In [10]:
ep_feat.citations()

['@article{ward_agrawal_choudary_wolverton_2016, title={A general-purpose machine learning framework for predicting properties of inorganic materials}, volume={2}, DOI={10.1038/npjcompumats.2017.28}, number={1}, journal={npj Computational Materials}, author={Ward, Logan and Agrawal, Ankit and Choudhary, Alok and Wolverton, Christopher}, year={2016}}']

## 2.2 Add more composition-based features

There are many more Composition based featurizers apart from ``ElementProperty`` that are available in the ``matminer.featurizers.composition``. Let's try the ``ElectronegativityDiff`` featurizer which requires knowing the oxidation state of the various elements in the Composition. This information is not there currently, but one can use the ``conversions`` package to try to guess oxidation states and then apply the ``ElectronegativityDiff`` featurizer to this column.

In [11]:
from matminer.featurizers.conversions import CompositionToOxidComposition
from matminer.featurizers.composition import OxidationStates

df = CompositionToOxidComposition().featurize_dataframe(df, "composition")

os_feat = OxidationStates()
df = os_feat.featurize_dataframe(df, "composition_oxid")
df.head()

CompositionToOxidComposition:   0%|          | 0/1181 [00:00<?, ?it/s]

OxidationStates:   0%|          | 0/1181 [00:00<?, ?it/s]

material_id    formula  space_group  \
0    mp-10003    Nb4CoSi          124   
1    mp-10010  Al(CoSi)2          164   
2    mp-10015       SiOs          221   
3    mp-10021         Ga           63   
4    mp-10025      SiRu2           62   

                                           structure  elastic_anisotropy  \
0  [[0.94814328 2.07280467 2.5112    ] Nb, [5.273...            0.030688   
1  [[0. 0. 0.] Al, [1.96639263 1.13529553 0.75278...            0.266910   
2   [[1.480346 1.480346 1.480346] Si, [0. 0. 0.] Os]            0.756489   
3  [[0.         1.09045794 0.84078375] Ga, [0.   ...            2.376805   
4  [[1.0094265  4.24771709 2.9955487 ] Si, [3.028...            0.196930   

        G_VRH       K_VRH  poisson_ratio   composition  \
0   97.141604  194.268884       0.285701  (Nb, Co, Si)   
1   96.252006  175.449907       0.268105  (Al, Co, Si)   
2  130.112955  295.077545       0.307780      (Si, Os)   
3   15.101901   49.130670       0.360593          (Ga)   
4  101.947798  256.768081       0.324682      (Si, Ru)   

   MagpieData minimum Number  ...  MagpieData maximum SpaceGroupNumber  \
0                       14.0  ...                                229.0   
1                       13.0  ...                                227.0   
2                       14.0  ...                                227.0   
3                       31.0  ...                                 64.0   
4                       14.0  ...                                227.0   

   MagpieData range SpaceGroupNumber  MagpieData mean SpaceGroupNumber  \
0                               35.0                        222.833333   
1                               33.0                        213.400000   
2                               33.0                        210.500000   
3                                0.0                         64.000000   
4                               33.0                        205.000000   

   MagpieData avg_dev SpaceGroupNumber  MagpieData mode SpaceGroupNumber  \
0                             9.611111                             229.0   
1                            15.520000                             194.0   
2                            16.500000                             194.0   
3                             0.000000                              64.0   
4                            14.666667                             194.0   

           composition_oxid  minimum oxidation state  maximum oxidation state  \
0        (Nb0+, Co0+, Si0+)                        0                        0   
1  (Al3+, Co2+, Co3+, Si4-)                       -4                        3   
2              (Si4-, Os4+)                       -4                        4   
3                    (Ga0+)                        0                        0   
4              (Si4-, Ru2+)                       -4                        2   

   range oxidation state  std_dev oxidation state  
0                      0                 0.000000  
1                      7                 3.872983  
2                      8                 5.656854  
3                      0                 0.000000  
4                      6                 4.242641  

[5 rows x 146 columns]

As you can see, the end of our data frame has now has some oxidation state based features!

## 2.3 Add some structure based features

Not all featurizers operate on compositions. Matminer can also analyze crystal structures and featurize those as well. Let's start by adding some simple density features.

In [12]:
from matminer.featurizers.structure import DensityFeatures

df_feat = DensityFeatures()
df = df_feat.featurize_dataframe(df, "structure")  # input the structure column to the featurizer
df.head()

DensityFeatures:   0%|          | 0/1181 [00:00<?, ?it/s]

material_id    formula  space_group  \
0    mp-10003    Nb4CoSi          124   
1    mp-10010  Al(CoSi)2          164   
2    mp-10015       SiOs          221   
3    mp-10021         Ga           63   
4    mp-10025      SiRu2           62   

                                           structure  elastic_anisotropy  \
0  [[0.94814328 2.07280467 2.5112    ] Nb, [5.273...            0.030688   
1  [[0. 0. 0.] Al, [1.96639263 1.13529553 0.75278...            0.266910   
2   [[1.480346 1.480346 1.480346] Si, [0. 0. 0.] Os]            0.756489   
3  [[0.         1.09045794 0.84078375] Ga, [0.   ...            2.376805   
4  [[1.0094265  4.24771709 2.9955487 ] Si, [3.028...            0.196930   

        G_VRH       K_VRH  poisson_ratio   composition  \
0   97.141604  194.268884       0.285701  (Nb, Co, Si)   
1   96.252006  175.449907       0.268105  (Al, Co, Si)   
2  130.112955  295.077545       0.307780      (Si, Os)   
3   15.101901   49.130670       0.360593          (Ga)   
4  101.947798  256.768081       0.324682      (Si, Ru)   

   MagpieData minimum Number  ...  MagpieData avg_dev SpaceGroupNumber  \
0                       14.0  ...                             9.611111   
1                       13.0  ...                            15.520000   
2                       14.0  ...                            16.500000   
3                       31.0  ...                             0.000000   
4                       14.0  ...                            14.666667   

   MagpieData mode SpaceGroupNumber          composition_oxid  \
0                             229.0        (Nb0+, Co0+, Si0+)   
1                             194.0  (Al3+, Co2+, Co3+, Si4-)   
2                             194.0              (Si4-, Os4+)   
3                              64.0                    (Ga0+)   
4                             194.0              (Si4-, Ru2+)   

   minimum oxidation state  maximum oxidation state  range oxidation state  \
0                        0                        0                      0   
1                       -4                        3                      7   
2                       -4                        4                      8   
3                        0                        0                      0   
4                       -4                        2                      6   

   std_dev oxidation state    density        vpa  packing fraction  
0                 0.000000   7.834556  16.201654          0.688834  
1                 3.872983   5.384968  12.397466          0.644386  
2                 5.656854  13.968635  12.976265          0.569426  
3                 0.000000   6.036267  19.180359          0.479802  
4                 4.242641   9.539514  13.358418          0.598395  

[5 rows x 149 columns]

In [13]:
df_feat.feature_labels()

['density', 'vpa', 'packing fraction']

Again, we see more features in the last few columns: ``density``, ``vpa``, and ``packing fraction``.

There are many more structure based featurizers that are much more complex and detailed analysis of the crystal structure that are outside of the scope of this example. Let's move on to doing some machine learning predictions.

# 3. Try some different machine learning models to relate input features to the bulk modulus

We now have enough data to do some machine learning! We'll need to first determine what columns we consider input (independent variables) and what column we consider output (dependent variable).

## 3.1 Define input data and output data

For now, we'll use ``K_VRH`` (bulk modulus) as the output. You could retry this example with ``G_VRH``, ``elastic_anisotropy``, or ``poisson_ratio`` as the target output.

For the inputs, we'll use all the features we generated. That is, everything except the output data and non-numerical columns like ``composition`` and ``structure``.



In [14]:
y = df['K_VRH'].values
excluded = ["G_VRH", "K_VRH", "elastic_anisotropy", "formula", "material_id",
            "poisson_ratio", "structure", "composition", "composition_oxid"]
X = df.drop(excluded, axis=1)
print("There are {} possible descriptors:\n\n{}".format(X.shape[1], X.columns.values))

There are 140 possible descriptors:

['space_group' 'MagpieData minimum Number' 'MagpieData maximum Number'
 'MagpieData range Number' 'MagpieData mean Number'
 'MagpieData avg_dev Number' 'MagpieData mode Number'
 'MagpieData minimum MendeleevNumber' 'MagpieData maximum MendeleevNumber'
 'MagpieData range MendeleevNumber' 'MagpieData mean MendeleevNumber'
 'MagpieData avg_dev MendeleevNumber' 'MagpieData mode MendeleevNumber'
 'MagpieData minimum AtomicWeight' 'MagpieData maximum AtomicWeight'
 'MagpieData range AtomicWeight' 'MagpieData mean AtomicWeight'
 'MagpieData avg_dev AtomicWeight' 'MagpieData mode AtomicWeight'
 'MagpieData minimum MeltingT' 'MagpieData maximum MeltingT'
 'MagpieData range MeltingT' 'MagpieData mean MeltingT'
 'MagpieData avg_dev MeltingT' 'MagpieData mode MeltingT'
 'MagpieData minimum Column' 'MagpieData maximum Column'
 'MagpieData range Column' 'MagpieData mean Column'
 'MagpieData avg_dev Column' 'MagpieData mode Column'
 'MagpieData minimum Row' 'Magpi

### 3.1 Try a linear regression model using scikit-learn

The scikit-learn library makes it easy to fit and cross-validate different types of regression models. Let's start with one of the simplest - a linear regression.

In [15]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

lr = LinearRegression()

lr.fit(X, y)

# get fit statistics
print('training R2 = ' + str(round(lr.score(X, y), 3)))
print('training RMSE = %.3f' % np.sqrt(mean_squared_error(y_true=y, y_pred=lr.predict(X))))

training R2 = 0.928
training RMSE = 19.611


In [16]:
print(X.shape, y.shape)

(1181, 140) (1181,)


This looks reasonable since linear regression is a simple (high bias) model. But to really validate that we are not over-fitting, we need to check the cross-validation score rather than the fitting score.

In [17]:
from sklearn.model_selection import KFold, cross_val_score

# Use 10-fold cross validation (90% training, 10% test)
crossvalidation = KFold(n_splits=10, shuffle=True, random_state=1)
scores = cross_val_score(lr, X, y, scoring='neg_mean_squared_error', cv=crossvalidation, n_jobs=1)
rmse_scores = [np.sqrt(abs(s)) for s in scores]
r2_scores = cross_val_score(lr, X, y, scoring='r2', cv=crossvalidation, n_jobs=1)

print('Cross-validation results:')
print('Folds: %i, mean R2: %.3f' % (len(scores), np.mean(np.abs(r2_scores))))
print('Folds: %i, mean RMSE: %.3f' % (len(scores), np.mean(np.abs(rmse_scores))))

Cross-validation results:
Folds: 10, mean R2: 0.902
Folds: 10, mean RMSE: 22.453


A root-mean-squared error of ~22 GPa is not bad! Let's see what this looks like on a plot.


In [18]:
import plotly.io as pio, plotly.graph_objects as go
from sklearn.model_selection import cross_val_predict

pio.renderers.default = "colab"   # use "png" (with kaleido) if interactive fails

# CV predictions with your linear regressor
y_pred = cross_val_predict(lr, X, y, cv=crossvalidation)

fig = go.Figure()
fig.add_scatter(x=y, y=y_pred, mode="markers", name="CV pred", text=df["formula"])
fig.add_scatter(x=[0, 400], y=[0, 400], mode="lines", name="y=x", line=dict(dash="dash"))

fig.update_layout(
    xaxis_title='DFT (MP) bulk modulus (GPa)',
    yaxis_title='Predicted bulk modulus (GPa)',
    title='Linear regression'
)

# keep equal scales so the y=x line is meaningful
fig.update_yaxes(scaleanchor="x", scaleratio=1)

fig.show()


Not too bad! However, there are definitely some outliers (you can hover over the points with your mouse to see what they are). We will see below (with a random forest model) how the accuracy changes when we try the model on a completely new test data.

## 3.2 Try a random forest model

Let's see if a more complex machine learning model does better. We can try a random forest model which is a good "starting" machine learning model, although one can usually do better. Let's repeat the steps for linear regression but for a random forest model.

In [19]:
from sklearn.ensemble import RandomForestRegressor

rf = RandomForestRegressor(n_estimators=50, random_state=1)

rf.fit(X, y)
print('training R2 = ' + str(round(rf.score(X, y), 3)))
print('training RMSE = %.3f' % np.sqrt(mean_squared_error(y_true=y, y_pred=rf.predict(X))))

training R2 = 0.989
training RMSE = 7.743


At least on the training data, we have very low RMSE and very high R2 - this is good! But let's see if these numbers hold up on cross-validation.

In [20]:
# compute cross validation scores for random forest model
r2_scores = cross_val_score(rf, X, y, scoring='r2', cv=crossvalidation, n_jobs=-1)
scores = cross_val_score(rf, X, y, scoring='neg_mean_squared_error', cv=crossvalidation, n_jobs=-1)
rmse_scores = [np.sqrt(abs(s)) for s in scores]

print('Cross-validation results:')
print('Folds: %i, mean R2: %.3f' % (len(scores), np.mean(np.abs(r2_scores))))
print('Folds: %i, mean RMSE: %.3f' % (len(scores), np.mean(np.abs(rmse_scores))))

Cross-validation results:
Folds: 10, mean R2: 0.924
Folds: 10, mean RMSE: 19.298


Looks like upon cross-validation, we do slightly better than a linear regression but not *too* much better. Let's plot this one as well.

In [21]:
import plotly.io as pio, plotly.graph_objects as go
pio.renderers.default = "colab"

y_pred = cross_val_predict(rf, X, y, cv=crossvalidation)

fig = go.Figure()
fig.add_scatter(x=y, y=y_pred, mode="markers", name="CV pred", text=df["formula"])
fig.add_scatter(x=[0,400], y=[0,400], mode="lines", name="y=x")
fig.update_layout(
    xaxis_title='DFT (MP) bulk modulus (GPa)',
    yaxis_title='Random forest bulk modulus (GPa)',
    title='Random forest regression'
)
fig.show()


Visually, this looks a little better but not *that* much better. The random forest did very well in training, but is only a little better than linear regression when we plot the cross-validation error as per above.

You could (optionally) visualize the *training* error by replacing the code ``cross_val_predict(rf, X, y, cv=crossvalidation)`` in the above cell with ``rf.predict(X)``. That would look a lot better, but would not be an accurate representation of your prediction error.

In [22]:
from sklearn.model_selection import train_test_split
X['formula'] = df['formula']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)
train_formula = X_train['formula']
X_train = X_train.drop('formula', axis=1)
test_formula = X_test['formula']
X_test = X_test.drop('formula', axis=1)

rf_reg = RandomForestRegressor(n_estimators=50, random_state=1)
rf_reg.fit(X_train, y_train)

# get fit statistics
print('training R2 = ' + str(round(rf_reg.score(X_train, y_train), 3)))
print('training RMSE = %.3f' % np.sqrt(mean_squared_error(y_true=y_train, y_pred=rf_reg.predict(X_train))))
print('test R2 = ' + str(round(rf_reg.score(X_test, y_test), 3)))
print('test RMSE = %.3f' % np.sqrt(mean_squared_error(y_true=y_test, y_pred=rf_reg.predict(X_test))))

training R2 = 0.988
training RMSE = 8.181
test R2 = 0.943
test RMSE = 16.717


In [23]:
import numpy as np
import plotly.graph_objects as go
import plotly.io as pio

pio.renderers.default = "colab"   # use "png" (with kaleido) if interactive fails

# Residuals
res_train = y_train - rf_reg.predict(X_train)
res_test  = y_test  - rf_reg.predict(X_test)

fig = go.Figure()

# Train histogram
fig.add_histogram(
    x=res_train,
    histnorm="probability",
    name="train",
    opacity=0.6,
    nbinsx=40
)

# Test histogram
fig.add_histogram(
    x=res_test,
    histnorm="probability",
    name="test",
    opacity=0.6,
    nbinsx=40
)

# Zero residual reference line
fig.add_shape(
    type="line", x0=0, x1=0, y0=0, y1=1, xref="x", yref="paper",
    line=dict(color="black", dash="dash")
)

fig.update_layout(
    title="Random forest regression residuals",
    xaxis_title="Bulk modulus prediction residual (GPa)",
    yaxis_title="Probability",
    barmode="overlay",
    showlegend=True
)

fig.show()



Finally, let's see what are the most important features used by the random forest model.

In [24]:
import numpy as np
import plotly.graph_objects as go
import plotly.io as pio

pio.renderers.default = "colab"   # use "png" (with kaleido) if interactive fails

# Feature importances from your trained RF
importances = rf.feature_importances_

# Get feature names robustly
try:
    feature_names = np.array(X.columns)
except Exception:
    feature_names = np.array([f"f{i}" for i in range(len(importances))])

# Sort and take top-k
idx = np.argsort(importances)[::-1]
k = 10
top_names = feature_names[idx][:k]
top_vals  = importances[idx][:k] * 100.0  # to %

# Bar chart
fig = go.Figure(
    go.Bar(x=top_names, y=top_vals, text=[f"{v:.2f}%" for v in top_vals], textposition="auto")
)
fig.update_layout(
    title="Feature importances (top 10)",
    xaxis_title="Feature",
    yaxis_title="Importance (%)",
    showlegend=False
)
fig.update_xaxes(tickangle=45)
fig.show()




Features relating to the melting point and to the volume per atom / density are the most important in the random forest model.

This concludes the tutorial! You are now familiar with some of the basic features of matminer.